Gsoc2020vivin

vivinperis opened the pull request:
Code changes

Commit:
kafka: Evaluate the topic name and query/insert into hashmap

Here, we evaluate the template topic name as and when a message arrives at the
Kafka destination.

The process starts with quering the hash map to check if the topic is there.
If it is not present, we insert the topic. Otherwise, we use the topic from
the hash map. The process is carried out under the protection of a mutex lock.

Signed-off-by: Vivin Peris <vivinperis98@gmail.com>

Commit:
kafka: Working prototype which takes care of valid topic names

The code here consists of expansion of topic name when it is a template,
checking if it is in hashmap. Insert if it isn't. The evt_tag_str have
been update to account for the presence of template in topic name.
Also create hash table using g_hash_table_new_full to benefit from the
destruction functions provided.

Signed-off-by: Vivin Peris <vivinperis98@gmail.com>

Commit:
kafka: Modified grammar for fallback topic and wrote validation logic

Made allowance for fallback topic which would help to prevent loss of messages
in case topic name is invalid. The driver ensures that the topic and fallback
topic are both valid
The flow is as follows
Case 1 - Topic name isn't a template.
Here, we check if the topic name is valid with conditions as in the official
kafka logic code. If the topic name isn't valid, the driver isn't initialized.
Here, we don't need fallback topic even though the grammar allows for the
definition of the same since its not possible for a message publish failure
due to a VALID topic name to be fixed by another valid topic name.

Case 2 - Topic name is a template.
Here, we first check if the topic name is in hashmap.
In case it isn't, we check if it is a valid topic name and if it is a
valid topic name, then insert into hashmap. Else, we use the fallback topic
if defined. In case fallback topic isn't defined, then we return false as it
is not possible to deliver the message.

Signed-off-by: Vivin Peris <vivinperis98@gmail.com>

bazsi commented on modules/kafka/kafka-dest-driver.c:
```diff
@@ -144,6 +153,27 @@ _kafka_log_callback(const rd_kafka_t *rkt, int level, const char *fac, const cha
 }
 
+rd_kafka_topic_t *
+_query_topic_from_hash_map(KafkaDestDriver *self, const gchar *name)
```
since this is a public function used by the worker, please use the kafka_dd_ prefix.

bazsi commented on modules/kafka/kafka-dest-worker.c:
```diff
@@ -61,30 +61,59 @@ _publish_message(KafkaDestWorker *self, LogMessage *msg)
+        {
+          if(_validate_topic_name(owner, topic_name_new->str)) //Insert into hashtable if topic name is valid
+            {
+              topic = _insert_topic_into_hash_map(owner, topic_name_new->str);
```
this code contains a race condition. you are locking the mutex in _query_topic_from_hash_map() function and if that doesn't return anything, you insert a new entry into the hash by locking it again.

what happens if two workers are working on the same topic? both would query and get a negative response and then go on about inserting the same topic into the hash.

vivinperis replied:
Thanks @bazsi. Missed this

bazsi commented on modules/kafka/kafka-dest-driver.c:
```diff
@@ -236,20 +266,73 @@ _construct_client(KafkaDestDriver *self)
+  gboolean is_callback = self->topicname_is_a_template && (!self->topic_hash_map);
+  if(len == 0)
+    {
+      msg_error("kafka: topic name is illegal, it can't be empty", is_callback ? evt_tag_str("Fallback topic", name) :
```
I don't think you need separate message tags for the fallback/normal case. If you do, please use identifier like tag names (e.g. `fallback_topic`, no spaces).

you can even remove the is_callback boolean if you let this go.

also, you can possible let the "self" argument go.

bazsi commented on modules/kafka/kafka-dest-driver.c:
```diff
@@ -236,20 +266,73 @@ _construct_client(KafkaDestDriver *self)
+    {
+      return rd_kafka_topic_new(self->kafka, name, NULL);
+    }
+  g_assert_not_reached();
```
I think it would be better to check rd_kafka_topic_new() returning NULL. If possible be better to make this a runtime error instead of an assert, even if that means we are dropping messages.

a crash with SIGABRT is not very user friendly, and although I am fan of using it in programmatic errors, we should only use g_assert() on those cases.

bazsi commented on modules/kafka/kafka-dest-driver.c:
```diff
@@ -277,7 +360,9 @@ _flush_inflight_messages(KafkaDestDriver *self)
-                 evt_tag_str("topic", self->topic_name),
+
+                 evt_tag_str("topic", self->topic_name->template),
+                 evt_tag_str("fallback_topic", (self->fallback_topic_name ? self->fallback_topic_name : "NULL")),
```
are you sure evt_tag_str() doesn't handle NULL values? I know it does :)

bazsi commented on modules/kafka/kafka-dest-driver.c:
```diff
@@ -287,7 +372,8 @@ _flush_inflight_messages(KafkaDestDriver *self)
       msg_error("kafka: error flushing accumulated messages during shutdown, rd_kafka_flush() returned failure, this might indicate that some in-flight messages are lost",
-                evt_tag_str("topic", self->topic_name),
+                evt_tag_str("topic", self->topic_name->template),
+                evt_tag_str("fallback_topic", (self->fallback_topic_name ? self->fallback_topic_name : "NULL")),
```
likewise (I am not repeating this any further).

bazsi commented on modules/kafka/kafka-dest-driver.c:
```diff
@@ -324,6 +410,29 @@ _purge_remaining_messages(KafkaDestDriver *self)
+  msg_debug("kafka: The topic name is a template");
+  if(!self->fallback_topic_name)
+    {
+      msg_error("kafka: Fallback topic required when the topic name is a template");
```
please use the correct spelling for the option, also using parentheses to indicate it is an actual option.

vivinperis replied:
I didn't understand this. Can you please elaborate?

bazsi replied:
In the error message. Let's assume the user doesn't have fallback-topic() in her config. In that case, she would receive the error message above. By spelling out the name of the configuration option precisely, she receives a lot more guidance.

bazsi commented on modules/kafka/kafka-dest-driver.c:
```diff
@@ -324,6 +410,29 @@ _purge_remaining_messages(KafkaDestDriver *self)
+  if(!self->fallback_topic_name)
+    {
+      msg_error("kafka: Fallback topic required when the topic name is a template");
+      g_assert_not_reached();
```
please handle this as a runtime error and not with an abort.

bazsi commented on modules/kafka/kafka-dest-driver.c:
```diff
@@ -324,6 +410,29 @@ _purge_remaining_messages(KafkaDestDriver *self)
+  if (self->fallback_topic == NULL)
+    {
+      msg_error("kafka: error constructing the fallback topic object",
+                evt_tag_str("fallback topic", self->fallback_topic_name),
```
underscore instead of space in evt_tag names.

bazsi commented on modules/kafka/kafka-dest-driver.c:
```diff
@@ -324,6 +410,29 @@ _purge_remaining_messages(KafkaDestDriver *self)
+                evt_tag_str("fallback topic", self->fallback_topic_name),
+                evt_tag_str("driver", self->super.super.super.id),
+                log_pipe_location_tag(&self->super.super.super.super));
+      g_assert_not_reached();
```
runtime error.

bazsi commented on modules/kafka/kafka-dest-driver.c:
```diff
@@ -333,34 +442,39 @@ kafka_dd_init(LogPipe *s)
-                evt_tag_str("driver", self->super.super.super.id),
-                log_pipe_location_tag(&self->super.super.super.super));
-      return FALSE;
+      _init_template_topic_name(self);
```
you can easily return the error condition here.

bazsi commented on modules/kafka/kafka-dest-driver.c:
```diff
@@ -333,34 +442,39 @@ kafka_dd_init(LogPipe *s)
                 log_pipe_location_tag(&self->super.super.super.super));
       return FALSE;
     }
-
```
I think a couple of empty lines help with readability, especially in such a long function.

bazsi commented on modules/kafka/kafka-dest-driver.h:
```diff
@@ -35,27 +35,35 @@ typedef struct
   GList *config;
   gchar *bootstrap_servers;
+  gchar *fallback_topic_name;
   rd_kafka_topic_t *topic;
```
is "topic" still needed?

vivinperis replied:
I was using it for the case where the topic name isn't a template in order to initialise it in the driver itself. Is there a way to get rid of it without increasing complexity(the other I could think of is store "Non template topic and fallback topic in the same topic)?

bazsi replied:
ok, understood.

bazsi commented on modules/kafka/kafka-dest-driver.h:
```diff
@@ -35,27 +35,35 @@ typedef struct
 void kafka_dd_merge_config(LogDriver *d, GList *props);
 void kafka_dd_set_bootstrap_servers(LogDriver *d, const gchar *bootstrap_servers);
-
+gboolean _validate_topic_name(KafkaDestDriver *self, const gchar *name);
```
this also seems to be a public function, use a proper prefix.

the underscore as a convention is used for private (static) functions.

bazsi commented on modules/kafka/kafka-dest-worker.c:
```diff
@@ -61,30 +61,59 @@ _publish_message(KafkaDestWorker *self, LogMessage *msg)
-
-  if (rd_kafka_produce(owner->topic,
+  rd_kafka_topic_t *topic;
+  GString *topic_name_new = g_string_sized_new(128);
```
this is a good candidate for scratch buffers which would avoid a per-message allocation.

bazsi commented on modules/kafka/kafka-dest-worker.c:
```diff
@@ -61,30 +61,59 @@ _publish_message(KafkaDestWorker *self, LogMessage *msg)
       msg_error("kafka: failed to publish message",
-                evt_tag_str("topic", owner->topic_name),
+                owner->topicname_is_a_template ? (!is_fallback_topic ? evt_tag_str("topic", topic_name_new->str) :
+                                                  evt_tag_str("fallback_topic", (owner->fallback_topic_name ? owner->fallback_topic_name : "NULL"))) :
```
I would just use the same tag name here with the actual topic we are publishing to. Either the output of the template or the fallback.

bazsi commented on modules/kafka/kafka-dest-worker.c:
```diff
@@ -61,30 +61,59 @@ _publish_message(KafkaDestWorker *self, LogMessage *msg)
-            evt_tag_str("topic", owner->topic_name),
+            owner->topicname_is_a_template ? (!is_fallback_topic ? evt_tag_str("topic", topic_name_new->str) :
+                                              evt_tag_str("fallback_topic", (owner->fallback_topic_name ? owner->fallback_topic_name : "NULL"))) :
+            evt_tag_str("topic", owner->topic_name->template),
```
same.

bazsi commented on modules/kafka/kafka-dest-worker.c:
```diff
@@ -121,7 +150,9 @@ _drain_responses(KafkaDestWorker *self)
-                evt_tag_str("topic", owner->topic_name),
+                owner->topicname_is_a_template ? evt_tag_str("template", owner->topic_name->template):
+                evt_tag_str("topic", owner->topic_name->template),
+                evt_tag_str("fallback topic", owner->fallback_topic_name ? owner->fallback_topic_name : "NULL"),
```
no space in fallback topic

bazsi reviewed (commented):
I left a number of comments, some of it more important than others.

MrAnno committed:
python: add PyBookmark

Signed-off-by: László Várady <laszlo.varady@protonmail.com>

MrAnno committed:
python: add unit test for PyBookmark

Signed-off-by: László Várady <laszlo.varady@protonmail.com>

MrAnno committed:
python: add python-ack-tracker

Signed-off-by: László Várady <laszlo.varady@protonmail.com>

ctr49 committed:
logmatcher: add dupnames pcre flag

Allow dupnames flag to be used in PCRE expressions, allowing duplicate names for named subpatterns as explained here https://www.pcre.org/original/doc/html/pcrepattern.html#SEC16

Signed-off-by: ctr49 <chris.ace@gmx.net>

MrAnno committed:
python: add LogMessage.set_bookmark()

Signed-off-by: László Várady <laszlo.varady@protonmail.com>

MrAnno committed:
python: initialize bookmark/ack-tracker in Python source tests

Signed-off-by: László Várady <laszlo.varady@protonmail.com>

MrAnno committed:
python: add InstantAckTrackerFactory

Signed-off-by: László Várady <laszlo.varady@protonmail.com>

Commit:
kafka: Added tests for the kafka topic parameter

Wrote unit tests for the _calculate_topic and _validate_topic_name methods to
 handle valid, and invalid topic names.
Created an internal header file for the functions to test and marked them
non static.

Refactored the kafka_dd_init method to take care of the case when topic name
is a literal. Also refactored the kafka_dd_validate_topic_name method to
remove dependency on the the driver object and modified the parametrized
test to take this into account.

Signed-off-by: Vivin Peris <vivinperis98@gmail.com>

Kokan committed:
filter-test: fix false positive check

Checking until the length is wrong, as the length is the length syslog-ng produces in case of 0 length it checks if the expected value equals to the computed value_msg for the first 0 chars, which is always true but if the expected (value) is not zero it should fail.

Signed-off-by: Kokan <kokaipeter@gmail.com>

Kokan committed:
filter/regex/test: extending current cases to cover all match

Signed-off-by: Kokan <kokaipeter@gmail.com>

Kokan committed:
filter/regex: fix not setting all of the named matches

The nametable which pcre provides can have gaps (begin_index=-1 || end_index=-1)
and in such cases those entries must be skipped and the next entry processed.
The issue was that the current implementation did only process the next name
entry if the current entry is not a gap. Witch works fine as long as there is no
named regex entry that is optional (see new UT cases for failing cases).

Signed-off-by: Kokan <kokaipeter@gmail.com>

Kokan committed:
news: add news file

Signed-off-by: Kokan <kokaipeter@gmail.com>

lbudai committed:
ack-tracker: add deinit method

Signed-off-by: Laszlo Budai <laszlo.budai@outlook.com>

lbudai committed:
ack-tracker: implement deinit method for BatchedAckTracker

Signed-off-by: Laszlo Budai <laszlo.budai@outlook.com>

lbudai committed:
ack-tracker: add init method

Signed-off-by: Laszlo Budai <laszlo.budai@outlook.com>

lbudai committed:
ack-tracker: implement init method for BatchedAckTracker

Signed-off-by: Laszlo Budai <laszlo.budai@outlook.com>

lbudai committed:
logsource: init/deinit AckTracker

Signed-off-by: Laszlo Budai <laszlo.budai@outlook.com>

lbudai committed:
ack-tracker: test BatchedAckTracker::deinit()

Signed-off-by: Laszlo Budai <laszlo.budai@outlook.com>

lbudai committed:
logthrsourcedrv: unref AckTrackerFactory

Signed-off-by: Laszlo Budai <laszlo.budai@outlook.com>

MrAnno committed:
python: add ConsecutiveAckTrackerFactory

Signed-off-by: László Várady <laszlo.varady@protonmail.com>

MrAnno committed:
python: add BatchedAckTrackerFactory

Signed-off-by: László Várady <laszlo.varady@protonmail.com>

MrAnno committed:
python: add unit test for ack tracker factories

Signed-off-by: László Várady <laszlo.varady@protonmail.com>

MrAnno committed:
python: use ack tracker factory in Python sources

The ack_tracker attribute can be used to request an ack tracker
instance in Python code. This attribute has to be set in the init()
method or earlier.

This sets an AckTrackerFactory, not a real ack tracker.

After the initialization of the source/fetcher, self.ack_tracker could
be overwritten with the real AckTracker instance in the future, if
needed.

For example:

from syslogng import LogSource
from syslogng import InstantAckTracker, ConsecutiveAckTracker, BatchedAckTracker

class TestSource(LogSource):
    def init(self, options):
        self.ack_tracker = InstantAckTracker(ack_callback=self.save_bookmark)
        self.ack_tracker = ConsecutiveAckTracker(ack_callback=self.save_bookmark)
        self.ack_tracker = BatchedAckTracker(timeout=5000, batch_size=100,
                                             batched_ack_callback=self.save_bookmarks)
        return True

    def save_bookmark(self, bookmark):
        pass

    def save_bookmarks(self, bookmarks):
        pass

    def run(self):
        pass

    def request_exit(self):
        pass

Signed-off-by: László Várady <laszlo.varady@protonmail.com>

MrAnno committed:
python: add bookmark and ack tracking support to Python sources

Example:

from syslogng import LogSource
from syslogng import LogMessage
from syslogng import InstantAckTracker, ConsecutiveAckTracker, BatchedAckTracker

import threading
import time

class TestSource(LogSource):
    def init(self, options):
        self.stop_event = threading.Event()

        self.ack_tracker = InstantAckTracker(ack_callback=self.save_bookmark)
        self.ack_tracker = ConsecutiveAckTracker(ack_callback=self.save_bookmark)
        self.ack_tracker = BatchedAckTracker(timeout=5000, batch_size=100,
                                             batched_ack_callback=self.save_bookmarks)
        return True

    def deinit(self):
        pass

    def save_bookmark(self, bookmark):
        print(bookmark)

    def save_bookmarks(self, bookmarks):
        for b in bookmarks:
            print(b)

    def run(self):
        while not self.stop_event.is_set():
            time.sleep(1)

            msg = LogMessage("test", bookmark="bookmark-data")
            # msg.set_bookmark("bookmark-data")
            self.post_message(msg)

    def request_exit(self):
        self.stop_event.set()

Signed-off-by: László Várady <laszlo.varady@protonmail.com>

MrAnno committed:
python: add optional constructor parameter "bookmark" to LogMessage

Signed-off-by: László Várady <laszlo.varady@protonmail.com>

Kokan committed:
version.sh: version number must be without newline

Signed-off-by: Kokan <kokaipeter@gmail.com>

MrAnno committed:
python: extract py_sd_get_ack_tracker()

Signed-off-by: László Várady <laszlo.varady@protonmail.com>

MrAnno committed:
Merge pull request #3396 from lbudai/batched-ack-tracker-ack-partial-batch

Batched ack tracker ack partial batch

MrAnno committed:
python: deinit ack tracker before running Python deinit()

Python source users may receive acknowledgements after they deinitialize
their source. This is true in general, but in case of BatchedAckTracker,
an incomplete but large batch will be acked after the deinitialization.

This patch reduces the number of post-deinit acks in case of
BatchedAckTracker, so users have a chance to react to those acks while
their source is still running.

Signed-off-by: László Várady <laszlo.varady@protonmail.com>

bazsi committed:
Merge pull request #3393 from Kokan/filter-regex-missing-named-entries

Filter regex missing named matches

bazsi committed:
Merge pull request #3399 from Kokan/scripts-version-no-line-end

version.sh: version number must be without newline

lbudai committed:
news: add news file for PR#3381

Signed-off-by: Laszlo Budai <laszlo.budai@outlook.com>

lbudai committed:
Merge pull request #3381 from ctr49/add-dupnames-pcre-flag

add dupnames pcre flag

furiel committed:
Merge pull request #3400 from lbudai/news-file-pcre-flag

news: add news file for PR#3381

lbudai committed:
Merge pull request #3392 from MrAnno/python-ack-bookmark

Python source acknowledgement

lbudai committed:
logpipe: when log_pipe_unref() calls free_fn, return TRUE

Signed-off-by: Laszlo Budai <laszlo.budai@outlook.com>

lbudai committed:
ack-tracker: fix LogSource lifetime in BatchedAckTracker

LogSource should be kept alive until we have
 * pending events (unfortunately when an event is posted, it will be
 called even after the event is unregistered)
 * running timers

Before posting the iv_ev_event, we should increase the refcount for
LogSource to guarantee that the LogSource is alive when the handler is
called.

Other part of the solution is to make sure that we have only one pending batch timer
restart requested event - otherwise we would increase the refcounter for
each full batches. This would be a problem in case of multiple pending
events, as the LogSource instance is kept alive only by these pending events and
the last timer restart event handler would free up the LogSource which means that
it would be a dangling pointer when the registered timer callback is called.

originally we had this in batch timer restart requested handler:
    if (!log_pipe_unref(&self->super.source->super))  ---> it won't true, due to other pending events
      _restart_batch_timer(self);

So if we have N batches that are full before the pending event is
executed, we would restart the timer N-1 times and the timer's cookie
would be a dangling pointer.

Signed-off-by: Laszlo Budai <laszlo.budai@outlook.com>

lbudai committed:
ack-tracker: add debug messages to BatchedAckTracker

Signed-off-by: Laszlo Budai <laszlo.budai@outlook.com>

lbudai committed:
ack-tracker: BatchedAckTracker: ack tracker is destroyed on its own thread

AckTracker::manage_msg_ack() is called from destination threads and when
LogSource refcnt reach 0, it's free() method will destroy its
AckTracker.
The problem is that the pending_request_restart_timer_lock is locked by
another thread - and we cannot call g_mutex_clear() in this case.
As this mutex can only be locked by the timer at ack_tracker_free()
time, we have to move the destruction of the AckTacker to the thread
that started the LogSource.

Signed-off-by: Laszlo Budai <laszlo.budai@outlook.com>

lbudai committed:
ack-tracker: BatchedAckTracker: register/unregister event from the owner thread

Signed-off-by: Laszlo Budai <laszlo.budai@outlook.com>

lbudai committed:
ack-tracker: BatchedAckTracker: move debug messages to trace level

Signed-off-by: Laszlo Budai <laszlo.budai@outlook.com>

lbudai committed:
Merge pull request #3406 from lbudai/fix-batched-ack-tracker-logsource-ref

Fix batched ack tracker logsource ref
